source/NVDAObjects/UIA/winConsoleUIA.py

```diff
@@ -31,6 +35,121 @@ def __init__(self, obj, position, _rangeObj=None):
 					1
 				)
 
+	def move(self, unit, direction, endPoint=None):
+		oldRange=None
+		if not self._isCaret:
+			# Insure we haven't gone beyond the visible text.
+			# UIA adds thousands of blank lines to the end of the console.
+			visiRanges = self.obj.UIATextPattern.GetVisibleRanges()
+			if visiRanges.length > 0:
+				firstVisiRange = visiRanges.GetElement(0)
+				lastVisiRange = visiRanges.GetElement(visiRanges.length - 1)
+				oldRange=self._rangeObj.clone()
+		if unit == textInfos.UNIT_WORD and direction != 0:
+			# UIA doesn't implement word movement, so we need to do it manually.
+			offset = self._getCurrentOffsetInThisLine()
+			index = 1 if direction > 0 else 0
+			start, end = self._getWordOffsetsInThisLine(offset)
+			wordMoveDirections = (
+				(offset - start) * -1,
+				end - offset
+			)
+			res = self.move(
+				textInfos.UNIT_CHARACTER,
+				wordMoveDirections[index],
+				endPoint=endPoint
+			)
+			if res != 0:
+				return direction
+			else:
+				if self.move(textInfos.UNIT_CHARACTER, -1): # Reset word boundaries to move to the previous word
+					return self.move(unit, direction, endPoint=endPoint)
+				else:
+					return res
+		res = super(consoleUIATextInfo, self).move(unit, direction, endPoint)
+		if oldRange and (
+			self._rangeObj.CompareEndPoints(
+				UIAHandler.TextPatternRangeEndpoint_Start,
+				firstVisiRange,
+				UIAHandler.TextPatternRangeEndpoint_Start
+			) < 0
+			or self._rangeObj.CompareEndPoints(
+				UIAHandler.TextPatternRangeEndpoint_Start,
+				lastVisiRange,
+				UIAHandler.TextPatternRangeEndpoint_End
+			) >= 0
+		):
+			self._rangeObj = oldRange
+			return 0
+		return res
+
+	def expand(self, unit):
+		if unit == textInfos.UNIT_WORD:
+			# UIA doesn't implement word movement, so we need to do it manually.
+			offset = self._getCurrentOffsetInThisLine()
+			start, end = self._getWordOffsetsInThisLine(offset)
+			wordEndPoints = (
+				(offset - start) * -1,
+				end - offset - 1
+			)
+			if wordEndPoints[0]:
+				self._rangeObj.MoveEndpointByUnit(
+					UIAHandler.TextPatternRangeEndpoint_Start,
+					UIAHandler.NVDAUnitsToUIAUnits[textInfos.UNIT_CHARACTER],
+					wordEndPoints[0]
+				)
+			if wordEndPoints[1]:
+				self._rangeObj.MoveEndpointByUnit(
+					UIAHandler.TextPatternRangeEndpoint_End,
+					UIAHandler.NVDAUnitsToUIAUnits[textInfos.UNIT_CHARACTER],
+					wordEndPoints[1]
+				)
+		else:
+			return super(consoleUIATextInfo, self).expand(unit)
+
+	def _getCurrentOffsetInThisLine(self):
+		lineInfo = self.copy()
+		lineInfo.expand(textInfos.UNIT_LINE)
+		charInfo = self.copy()
+		res = 0
+		chars = None
+		while True:
```

Comment:
Recent experiences with Notepad++ made me a bit worried about endless while loops. Could you limit this somehow?

Would something like this work?

```suggestion
		while charInfo.compareEndPoints(
			lineInfo,
			"startToEnd"
		) <= 0:
```